File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/transaction/GLAccountBalanceItem.java
package com.jasonzqshen.familyaccounting.core.transaction;

import java.util.Hashtable;

import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity_GLAccount;
import com.jasonzqshen.familyaccounting.core.utils.CurrencyAmount;

public class GLAccountBalanceItem {
	private final Hashtable<MonthIdentity, CurrencyAmount> _list;
	private final MasterDataIdentity_GLAccount _glAccount;
	private CurrencyAmount _sum = new CurrencyAmount();

	GLAccountBalanceItem(MasterDataIdentity_GLAccount glAccount) {
		_glAccount = glAccount;
		_list = new Hashtable<MonthIdentity, CurrencyAmount>();

	}

	/**
	 * get G/L account
	 * 
	 * @return
	 */
	public MasterDataIdentity_GLAccount getGLAccount() {
		return _glAccount;
	}

	/**
	 * add value
	 * 
	 * @param monthId
	 *            month identity
	 * @param value
	 */
	void addAmount(MonthIdentity monthId, CurrencyAmount amount) {
		_sum.addTo(amount);

		if (_list.contains(monthId)) {
			CurrencyAmount sum = _list.get(monthId);
			sum.addTo(amount);
		} else {
			_list.put(monthId, new CurrencyAmount(amount));
		}
	}

	/**
	 * get G/L account balance in
	 * 
	 * @param startId
	 * @return
	 */
	public CurrencyAmount getSumAmount(MonthIdentity startId,
			MonthIdentity endId) {
		CurrencyAmount sum = new CurrencyAmount();
		for (MonthIdentity id : _list.keySet()) {
			if (id.compareTo(startId) >= 0 && id.compareTo(endId) <= 0) {
				sum.addTo(_list.get(id));
			}
		}

		return sum;
	}

	/**
	 * get G/L account sum
	 * 
	 * @return
	 */
	public CurrencyAmount getSumAmount() {
		return new CurrencyAmount(_sum);
	}

	/**
	 * get amount for current month
	 * 
	 * @param id
	 * @return
	 */
	public CurrencyAmount getAmount(MonthIdentity id) {
		if (_list.contains(id)) {
			CurrencyAmount amount = _list.get(id);
			return new CurrencyAmount(amount);
		}
		return new CurrencyAmount();
	}
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Transaction/GLAccountBalanceItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Transaction
{
    public class GLAccountBalanceItem
    {
        private readonly Dictionary<MonthIdentity, CurrencyAmount> _list;
        private readonly MasterDataIdentity_GLAccount _glAccount;
        public MasterDataIdentity_GLAccount GLAccount { get { return _glAccount; } }

        private CurrencyAmount _sum;
        public CurrencyAmount Sum { get { return new CurrencyAmount(_sum); } }


        internal GLAccountBalanceItem(MasterDataIdentity_GLAccount glAccount)
        {
            _glAccount = glAccount;
            _list = new Dictionary<MonthIdentity, CurrencyAmount>();
            _sum = new CurrencyAmount();

        }

        /// <summary>
        /// Add amount 
        /// </summary>
        /// <param name="monthId"></param>
        /// <param name="amount"></param>
        internal void AddAmount(MonthIdentity monthId, CurrencyAmount amount)
        {
            _sum.AddTo(amount);

            if (_list.ContainsKey(monthId))
            {
                CurrencyAmount sum;
                _list.TryGetValue(monthId, out sum);
                sum.AddTo(amount);
            }
            else
            {
                _list.Add(monthId, new CurrencyAmount(amount));
            }
        }

        /// <summary>
        /// Get sum of amount from start month to end month
        /// </summary>
        /// <param name="startId">id of start month</param>
        /// <param name="endId">id of end month</param>
        /// <returns></returns>
        public CurrencyAmount GetSumAmount(MonthIdentity startId,
                MonthIdentity endId)
        {
            CurrencyAmount sum = new CurrencyAmount();
            foreach (var item in _list)
            {
                if (item.Key.CompareTo(startId) >= 0 && item.Key.CompareTo(endId) <= 0)
                {
                    sum.AddTo(item.Value);
                }
            }

            return sum;
        }

        /// <summary>
        /// Get amount
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        public CurrencyAmount GetAmount(MonthIdentity id)
        {
            if (_list.ContainsKey(id))
            {
                CurrencyAmount amount;
                _list.TryGetValue(id, out amount);
                return new CurrencyAmount(amount);
            }
            return new CurrencyAmount();
        }
    }

}
